### PR TITLE
Added the FLAMEGUY death animation

### PR DESCRIPTION
**Pull request by Mailaender**
A revival of https://github.com/OpenRA/OpenRA/pull/9350.

### CODE REVIEW

File: OpenRA.Mods.Common/Lint/CheckDeathTypes.cs [hunk 1]
```diff
@@ -51,7 +55,7 @@ public void Run(Action<string> emitError, Action<string> emitWarning, Ruleset ru
 							continue;
 
 						if (!warhead.DamageTypes.Overlaps(deathTypes))
-							emitError("Actor type `{0}` does not define a death animation for weapon `{1}`!"
+							emitError("Actor type `{0}` does neither define a death animation nor spawns an actor on death for weapon `{1}`!"
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckDeathTypes.cs]: Weird english here.  "Actor type `{0}` doesn't define a death animation or spawn an actor on death for weapon `{1}`!" would be better.

**Comment by pchote** [comment]
The running speed feels comically fast to me.  Can you slow it down?

Looks good aside from those nits, so :+1: after they are fixed.

**Comment by Mailaender** [comment]
Slowed it down.

File: OpenRA.Mods.Common/Lint/CheckDeathTypes.cs [hunk 2]
```diff
@@ -36,6 +36,10 @@ public void Run(Action<string> emitError, Action<string> emitWarning, Ruleset ru
 				if (!targetable.Any())
 					continue;
 
+				var sa = actorInfo.Value.TraitInfos<SpawnActorOnDeathInfo>();
+				if (sa.Any())
+					deathTypes.AddRange(sa.Where(t => t.DeathType != null).Select(dt => dt.DeathType).ToList());
```
Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckDeathTypes.cs]: I believe the `ToList` here is redundant.

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckDeathTypes.cs]: This whole block can simplify down to something like:
`deathTypes.AddRange(actorInfo.Value.TraitInfos<SpawnActorOnDeathInfo>().SelectMany(s => s.DeathType).Where(dt => dt != null).Distinct()))`

Comment (pchote) [inline OpenRA.Mods.Common/Lint/CheckDeathTypes.cs]: This code won't run at all if the actor doesn't have any death type animations.  You will need to adjust the checks above to account for `SpawnActorOnDeath`.

**Comment by pchote** [comment]
Removing my original :+1: because I didn't properly test the lint rules.

**Comment by Marnador** [comment]
From my test the flameguy shadow sequence seems a bit off. 
My suggestion is to put  **ShadowStart: 148** under **Defaults:** and remove **ShadowStart: 48** under **run:**. That way it doesn't flickers on and off when the actor is standing still for a few milliseconds, and it will render the right shadows when the actor is running.
But otherwise works pretty well! was actually looking in to it not too long ago :)

**Comment by Mailaender** [comment]
Rewrote the lint check and fixed the shadow sequence.

File: mods/ts/sequences/infantry.yaml [hunk 3]
```diff
@@ -1706,9 +1622,21 @@ civ3:
...
-	die5: flameguy # TODO: walking animation unused
-		Start: 42
-		Length: 104
-		ShadowStart: 190
 	die6: electro
 		Length: *
+
+flameguy:
+	Defaults:
+		Facings: 7
```
Comment (reaperrr) [inline mods/ts/sequences/infantry.yaml]: This looks highly suspicious. Does the art really cover only 7 facings?

Comment (Mailaender) [inline mods/ts/sequences/infantry.yaml]: No, it is wrong: `8` like all the others.

Comment (reaperrr) [inline mods/ts/sequences/infantry.yaml]: Nevermind, it actually does.

Comment (reaperrr) [inline mods/ts/sequences/infantry.yaml]: * sigh* scratch that, I somehow skipped the frames of the 8th facing.

**Comment by pchote** [comment]
One last glitch: I don't think this should apply to cyborgs.  It looks rather strange for a broken robot to suddenly grow legs and run around on fire.

**Comment by Mailaender** [comment]
This should now also take cyborgs into account.

**Comment by pchote** [comment]
Cyborgs still look quite glitchy, instantly growing new legs before falling over again on fire.

**Comment by reaperrr** [comment]
In my opinion cyborgs should just explode on death (regardless of death type) like they did in the original, the WithInjury logic makes anything else look glitchy without custom art.

**Comment by Mailaender** [comment]
Explosible cyborgs seems unrelated to this pull request (it doesn't regress it) and I want to keep the proper setup in the commit history.

**Comment by pchote** [comment]
The tag is there for reverting the specific flame death additions for cyborgs, so that they remain as they currently are on bleed.

**Comment by Mailaender** [comment]
They already remain as they currently are on bleed.

**Comment by pchote** [comment]
ab415d874a835bd97289ecd91f14c2c769ba306d adds a `FLAMECYBORG` actor and sequences.

**Comment by Mailaender** [comment]
It isn't used at the moment, but that doesn't mean the sequence setup should be thrown away.

**Comment by pchote** [comment]
Do you have a usecase for it?

**Comment by reaperrr** [comment]
> # TODO: unused running frame 322 and following

That comment was simply wrong (at least in the sense that it implied they should be used), `die5` needs to be a duplicate of `die4` for cyborgs. Actually, the whole Cyborg death setup is a bit bogus currently. Fixing that is outside the scope of this PR of course, but that's just another reason to only concentrate on avoiding changes/regressions over bleed.

> The obvious use case for FLAMECYBORG is using it for effects in maps and missions or triggering the cyborg crippling only for certain weapons or at random.

That goes against our "make TS as close to the original first before we make custom changes" policy. And mappers can still do that via custom map rules, if they want to.

**Comment by Mailaender** [comment]
Added the explosion back in.

**Comment by reaperrr** [comment]
@pchote's (and mine) fixup request for not adding unused sequences etc. still stands.

**Comment by Mailaender** [comment]
Removed the `FLAMECYBORG`. At least the past commit is referenced so not all is wasted.

**Comment by pchote** [comment]
The cyborg explosion is now drawn in the wrong palette.  I suspect you are missing a palette override on the `WithDeathAnimation`.

**Comment by Mailaender** [comment]
Updated again. I would have wished if all this unrelated crap could have stayed out of this pull request.

**Comment by reaperrr** [comment]
Looks good to me now :+1:

**Comment by pchote** [comment]
:+1:

**Comment by abcdefg30** [comment]
Thanks for reviving this, @Mailaender!

**Pull request merged**